### DISCUSSION TITLE
Self-contained deployment not working? - MonoGame w/ NetBeauty

### DISCUSSION THREAD

**Discussion by lunkums** [nulastudio/NetCoreBeauty#41] [Q&A] [answered]
Howdy,

I'm trying to deploy my [MonoGame ](https://docs.monogame.net/articles/packaging_games.html) project as a standalone application with Netbeauty. I run the following command in the .NET CLI:

> dotnet publish -c Release -r win-x64 /p:PublishReadyToRun=false /p:TieredCompilation=false --self-contained

When I have:

> &lt;DisableBeauty&gt;True&lt;/DisableBeauty&gt;

in my .csproj file, the deployed application is ~ 145mb. When I have:

> &lt;DisableBeauty&gt;False&lt;/DisableBeauty&gt;

in my .csproj file, the deployed application is ~70mb. This leads me to believe NetBeauty is excluding the .NET runtime from the deployment. Here are the contents of my .csproj file: https://pastebin.com/xRp8XzVr

Is it true that NetBeauty is excluding the .NET runtime from my deployed game, or is it automatically cleaning up unused files?

Thanks,
Luke

**Comment by liesauer** [comment]
no, NetBeauty doesn't remove any references. it's kind of weird. can i get the both copies of the deployed application with beauty enable and beauty diable? the files structure is ok too(you can get the structure with `tree` command), but it's better that have the application copies for furture testing.

**Reply by lunkums** [reply]
Yup, here's the two deployed applications as .zip files: https://drive.google.com/drive/folders/1MzLVi_9BDv3oGNMzI9Xr6bwC95XX3zIw?usp=sharing

And here's the project itself on GitHub: https://github.com/lunkums/LD51

**Reply by Starciad** [reply]
I encountered the same issue and found that, in MonoGame DesktopGL projects, certain native libraries must remain in the root directory where the executable is located. Because these files are loaded dynamically, they should be added to the ignore section of NetBeauty to prevent them from being relocated.

In most cases, the following tag is sufficient to exclude all relevant files:

```xml
<BeautyExcludes>SDL2.dll;libSDL2-2.0.so.0;libopenal.so.1;libopenal.1.dylib;libSDL2.dylib;soft_oal.dll</BeautyExcludes>
```

MonoGame’s loading mechanism searches for these libraries in the same directory as the running `.exe`. If your game still reports missing dependencies, add a try-catch block around the construction of your Game-derived class to identify which libraries are failing to load.

For WindowsDX projects, I did not encounter any DLL loading issues.

**Comment by liesauer** [comment, marked as answer]
i found out what happened.
the first time you published the application with NetBeauty enable, all the assemblies have been move to `libraries` which is expected.
the second time you published the application with NetBeauty disable, msbuild doesn't clean up the `publish` folder and leaves the old `libraries` in there, that is why you get double size of the assemblies.

**Comment by liesauer** [comment]
you should clean up the `publish` folder when switching the config by simply delete it.

**Reply by lunkums** [reply]
doh. You're right. I usually do this, but I think I was in a rush and figured there was some other problem. Thank you for your help :)

**Reply by liesauer** [reply]
it's glad that help.
